File: Temp/Proc-0/Machine/Amd64/Amd64Reader.cs
/// <summary>
/// Лексический анализатор двоичного кода архитектуры AMD64.
/// </summary>
class Amd64Reader
{
	private CharReader _reader;

	private FileStream _writer;
	private Utf8 _utf8;

	public Amd64Reader(CharReader reader)
	{
		_reader = reader;

		//_writer = new FileStream("debug.txt", O_CREAT | O_WRONLY | O_BINARY, _S_IREAD | _S_IWRITE);
		//_utf8 = new Utf8((Writer*)_writer);
	}

	//~Amd64Reader()
	//{
	//	delete _utf8;
	//	_writer->Close();
	//	delete _writer;
	//}

	public bool Get()
	{
		int symbol = _reader.ReadChar();
		if (symbol > 0)
		{
			// Вывод UTF-8.
			if (symbol < 256)
				Console.Write("%c", symbol);
			else
				Console.Write("[%u]", symbol);
			// Запись UTF-8.
			_utf8.WriteChar(symbol);

			return true;
		}

		return false;
	}
}

File: Proc-0/src/main/kotlin/Machine/Amd64/Amd64Reader.kt
package proc.Machine.Amd64

import proc.Interop.FileWriter
import proc.Text.CharReader
import proc.Text.Utf8

/**
 * Лексический анализатор двоичного кода архитектуры AMD64.
 */
class Amd64Reader {

	private var _reader: CharReader

	private var _writer: FileWriter
	//private var _utf8: Utf8

	init {
	}

	constructor(reader: CharReader) {
		_reader = reader
		_writer = FileWriter("debug.txt")
		//_utf8 = Utf8(null, _writer)
	}

	// TODO Avoid this.
	//protected fun finalize() {
	//	delete _utf8
	//	_writer->Close()
	//	delete _writer
	//}

	fun Get(): Boolean {
		var symbol: Int = _reader.ReadChar()
		if (symbol > 0) {

			// Вывод UTF-8.
			if (symbol < 256)
				print("$symbol")
			else
				print("[$symbol]")

			// Запись UTF-8.
			//_utf8.WriteChar(symbol)

			return true
		}

		return false
	}
}
